File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/sql-provenance.py
#!/usr/bin/env python3
"""Write or verify Custom SQL element provenance."""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent / "lib"))
import sql_provenance  # noqa: E402


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--workdir", required=True)
    parser.add_argument("--dm-spec")
    parser.add_argument("--check", action="store_true")
    args = parser.parse_args()
    workdir = Path(args.workdir).expanduser().resolve()
    dm_spec = (
        Path(args.dm_spec).resolve()
        if args.dm_spec
        else next(
            (
                path
                for path in (
                    workdir / "dm-remapped.json",
                    workdir / "dm-raw.json",
                    workdir / "dm-spec.json",
                )
                if path.is_file()
            ),
            None,
        )
    )
    artifact = workdir / "sql-provenance.json"
    if dm_spec is None or not dm_spec.is_file():
        print("SQL PROVENANCE FAIL: no data-model spec is available", file=sys.stderr)
        return 3
    try:
        result = sql_provenance.evaluate(
            dm_spec,
            workdir / "conv-meta.json",
            workdir / "workbook-content.twb",
            workdir / "custom-sql.json",
            workdir / "sql-provenance-overrides.json",
        )
        if args.check:
            actual = json.loads(artifact.read_text(encoding="utf-8-sig"))
            if actual != result:
                print(
                    "SQL PROVENANCE FAIL: sql-provenance.json is stale or was edited",
                    file=sys.stderr,
                )
                return 3
        else:
            artifact.write_text(
                json.dumps(result, indent=2, ensure_ascii=False) + "\n",
                encoding="utf-8",
            )
    except (OSError, ValueError, json.JSONDecodeError) as exc:
        print(f"SQL PROVENANCE FAIL: {exc}", file=sys.stderr)
        return 3
    print(
        f"SQL provenance: {result['status'].upper()} — "
        f"{len(result['sql_elements'])} SQL element(s), "
        f"{len(result['blockers'])} unattributed"
    )
    if result["status"] == "fail":
        for entry in result["blockers"]:
            print(
                f"  - {entry.get('element_name') or entry.get('element_id')}: "
                f"{entry['evidence']}",
                file=sys.stderr,
            )
        return 3
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/sql-provenance.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'optparse'
require_relative 'lib/sql_provenance'

opts = {}
OptionParser.new do |parser|
  parser.on('--workdir DIR') { |value| opts[:workdir] = File.expand_path(value) }
  parser.on('--dm-spec PATH') { |value| opts[:dm_spec] = File.expand_path(value) }
  parser.on('--check') { opts[:check] = true }
end.parse!

workdir = opts[:workdir] || abort('usage: sql-provenance.rb --workdir DIR [--dm-spec PATH] [--check]')
dm_spec = opts[:dm_spec] ||
          [File.join(workdir, 'dm-spec.json'), File.join(workdir, 'dm-raw.json')]
          .find { |path| File.file?(path) }
artifact = File.join(workdir, 'sql-provenance.json')
unless dm_spec && File.file?(dm_spec)
  warn 'SQL PROVENANCE FAIL: no data-model spec is available'
  exit 3
end

result = begin
  SqlProvenance.evaluate(
    dm_spec_path: dm_spec,
    metadata_path: File.join(workdir, 'conv-meta.json'),
    twb_path: File.join(workdir, 'workbook-content.twb'),
    custom_sql_path: File.join(workdir, 'custom-sql.json'),
    overrides_path: File.join(workdir, 'sql-provenance-overrides.json')
  )
rescue JSON::ParserError, SystemCallError => e
  warn "SQL PROVENANCE FAIL: #{e.message}"
  exit 3
end

if opts[:check]
  actual = begin
    JSON.parse(File.read(artifact))
  rescue JSON::ParserError, SystemCallError => e
    warn "SQL PROVENANCE FAIL: #{artifact} is missing/unreadable: #{e.message}"
    exit 3
  end
  unless actual == result
    warn 'SQL PROVENANCE FAIL: sql-provenance.json is stale or was edited; rerun pass 1'
    exit 3
  end
else
  File.write(artifact, "#{JSON.pretty_generate(result)}\n")
end

puts "SQL provenance: #{result['status'].upcase} — #{result['sql_elements'].size} SQL element(s), " \
     "#{result['blockers'].size} unattributed"
if result['status'] == 'fail'
  result['blockers'].each do |entry|
    warn "  - #{entry['element_name'] || entry['element_id']}: #{entry['evidence']}"
  end
  warn "Add a reasoned entry with a `proof` JSON path (`match:true`) to " \
       "#{File.join(workdir, 'sql-provenance-overrides.json')} only after proving its source semantics."
  exit 3
end
exit 0
